specs/fault-proof.md

```diff
@@ -0,0 +1,412 @@
+
+`128` specifically (`1000 0000` in binary) is reserved for key-type length-extension
+(reducing the content part to `30` or less key bytes), if the need arises.
+
+#### Type `129-255`: application usage
+
+This range of key types may be used by forks or customized versions of the fault proof protocol.
+
+### Bootstrapping
+
+Initial inputs are deterministic, but not necessarily singular or global:
+there may be multiple different disputes at the same time, each with its own disputed claims and L1 context.
+
+To bootstrap, the program requests the initial inputs from the VM, using pre-image key type `1`.
+
+The VM is aware of the external context, and maps requested pre-image keys based on their type, i.e.
+a local lookup for type `1`, or global one for `2`, and optionally support other key-types.
+
+### Hinting
+
+There is one more form of optional communication between client and server: pre-image hinting.
+Hinting is optional, and *is a no-op* in a L1 VM implementation.
+
+The hint itself comes at very low cost onchain: the hint can be a single `write` sys-call,
+which is instant as the memory to write as hint does not actually need to be loaded as part of the onchain proof.
+
+Hinting allows the program, when generating a proof offchain,
+to instruct the VM what data it is interested in.
+
+The VM can choose to execute the requested hint at any time: either locally (for standard requests),
+or in a modular form by redirecting the hint to tooling that may come with the VM program.
+
+Hints do not have to be executed directly: they may first just be logged to show the intents of the program,
+and the latest hint may be buffered for lazy execution, or dropped entirely when in read-only mode (like onchain).
+
+When the pre-image oracle serves a request, and the request cannot be served from an existing collection of pre-images
+(e.g. a local pre-image cache) then the VM can execute the hint to retrieve the missing pre-image(s).
+It is the responsiblity of the program to provide sufficient hinting for every pre-image request.
+Some hints may have to be repeated: the VM only has to execute the last hint when handling a missing pre-image.
+
+Note that hints may produce multiple pre-images:
+e.g. a hint for an ethereum block with transaction list may prepare pre-images for the header,
+each of the transactions, and the intermediate merkle-nodes that form the transactions-list Merkle Patricia Trie.
+
+Hinting is implemented with a minimal wire-protocol over a blocking one-way stream:
+
+```text
+<request> := <length prefix> <hint bytes> <end>
+<length prefix> := big-endian uint32  # length of <hint bytes>
+<hint bytes> := byte sequence
+<end> := 0 byte
+```
+
+The `<end>` trailing zero byte allows the server to block the program
+(since the communication is blocking) until the hint is processed.
+
+### Pre-image communication
+
+Pre-images are communicated with a minimal wire-protocol over a blocking two-way stream:
```

Comment:
Why "wire-protocol over a blocking two-way stream"? A blocking syscall seems to describe this better. More specifically this seems to pull from the unix syscall lineage with the following API

```
// Header
int open(char *prefixed_pre_image_hash)
int read(int fd, char *buf)


// Usage
fd = open(prefixed pre-image hash)
buf = malloc(...)
while(true) {
    c = read(fd, &buf);
    if (c == 0) break;
}

```

Reply:
Yes, it's meant to be implemented with the syscall interface. But I didn't want to enshrine linux syscalls in this spec, although I do think it works very well: we get to avoid magic memory-mapping, and can unify onchain and offchain pre-image oracle code-paths. Since offchain, even outside of a VM emulator, you can just attach a file-descriptor to the program process to serve it the pre-images.
I'll add the blocking syscall as example of this behavior outside of the pre-image oracle thing.

Reply:
I mean we should define our own types & condition rather than exactly copying linux here, but I think we should define the write/read/pipe syscalls (though maybe not now) because that will be how the program interacts with the world (and then you need to layer the wire format on top of the syscalls)

Reply:
Definitely, I can add the specific syscall pre-image oracle usage in a separate specs PR, I'll track it in linear. On a related note I need to specify exactly which things the program may use at all; Go itself uses syscalls as well. This crosses over in the FPVM project though, I'll create a linked issue there.